File: day01/Main.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Scanner;

public class Main {
    static ArrayList<ArrayList<Integer>> readFile() {
        ArrayList<Integer> left = new ArrayList<Integer>();
        ArrayList<Integer> right = new ArrayList<Integer>();
        File myFile = new File("input.txt");
        try {
            Scanner scanner = new Scanner(myFile);
            while (scanner.hasNextLine()) {
                String data = scanner.nextLine();
                String[] elements = data.split(" +");
                left.add(Integer.parseInt(elements[0]));
                right.add(Integer.parseInt(elements[1]));
            }
            scanner.close();
        } catch (FileNotFoundException e) {
            System.err.println(e);
        }
        ArrayList<ArrayList<Integer>> output = new ArrayList<ArrayList<Integer>>();
        output.add(left);
        output.add(right);
        return output;
    }

    static void task1() {
        ArrayList<ArrayList<Integer>> columns = readFile();
        ArrayList<Integer> left = columns.get(0);
        ArrayList<Integer> right = columns.get(1);
        Collections.sort(left);
        Collections.sort(right);
        int sum = 0;
        for (int i=0; i < left.size(); i++) {
            sum += Math.abs(left.get(i) - right.get(i));
        }
        System.out.println(sum);
    }

    static void task2() {
        ArrayList<ArrayList<Integer>> columns = readFile();
        ArrayList<Integer> left = columns.get(0);
        ArrayList<Integer> right = columns.get(1);
        HashMap<Integer, Integer> counter = new HashMap<>();
        for (Integer element: right) {
            Integer value = counter.get(element);
            if (value == null) {
                value = 0;
            }
            counter.put(element, value + 1);
        }
        int sum = 0;
        for (Integer element: left) {
            Integer value = counter.get(element);
            if (value == null) {
                continue;
            }
            sum += value * element;
        }
        System.out.println(sum);
    }

    public static void main(String[] args) {
        task1();
        task2();
    }
}

File: day01/main.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <math.h>
#include <unordered_map>

std::vector<std::vector<int> > readFile() {
    std::ifstream inputFile;
    inputFile.open("input.txt");
    std::string text;
    std::vector<int> left;
    std::vector<int> right;
    while ( inputFile.is_open() ) {
        inputFile >> text;
        left.insert(left.begin(), std::stoi(text));
        inputFile >> text;
        right.insert(right.begin(), std::stoi(text));
        if (inputFile.eof()) {
            break;
        }
    }
    inputFile.close();
    std::vector<std::vector<int> > output;
    output.insert(output.begin(), left);
    output.insert(output.begin(), right);
    return output;
}

void task1() {
    std::vector<std::vector<int> > data = readFile();
    std::vector<int> left = data.at(0);
    std::vector<int> right = data.at(1);
    sort(left.begin(), left.end());
    sort(right.begin(), right.end());
    int sum = 0;
    for (int i=0; i<left.size(); i++) {
        sum += abs(left.at(i) - right.at(i));
    }
    std::cout << sum << std::endl;
}

void task2() {
    std::vector<std::vector<int> > data = readFile();
    std::vector<int> left = data.at(0);
    std::vector<int> right = data.at(1);
    std::unordered_map<int, int> counter;
    int times;
    int element;
    for (int i=0; i<right.size(); i++) {
        element = right.at(i);
        std::unordered_map<int,int>::const_iterator found = counter.find(element);
        if (found == counter.end()) {
            counter[element] = 1;
        } else {
            counter[element] = found->second + 1;
        }
    }
    int sum = 0;
    for (int i=0; i<left.size(); i++) {
        element = left.at(i);
        std::unordered_map<int,int>::const_iterator found = counter.find(element);
        if (found != counter.end()) {
            sum += element * found->second;
        }
    }
    std::cout << sum << std::endl;
}

int main() {
    task1();
    task2();
}